keps/sig-instrumentation/20191115-structured-logging.md

```diff
@@ -0,0 +1,424 @@
+
+#### Alpha
+
+Preparation and migration of selected logs to new logging methods:
+* Support for the new InfoS, WarningS, ErrorS and FatalS method will be implemented in klog and the new version of klog will be released.
+* Kubernetes will be upgraded to use the new klog version.
+* Helper methods for producing consistent Kubernetes resource identifiers to be used in logs will be implemented.
+* Most important klog calls will be identified and will be manually migrated from using format based to structured methods.
+
+#### Beta
+
+Adding support for JSON output format and guarding against regression:
+* klog version used by Kubernetes and all of its dependencies will be upgraded to v2.
+* Implementation of logr which will produce logs in JSON format will be created.
+* Flag which will inject new logr implementation into klog if --logging-format is set to json will be added to k8s.io/component-base JSON format
+* Static analysis protects against adding new string formatting logs
+
+#### GA
+
+* All feedback for text and json logs output formats is addressed and both output formats become an API.
+
+### Performance
+
+Proposed changes could impact k8s performance in two ways:
+* By introducing degrading in klog implementation
+* By largely increasing log volume
+
+#### Logger implementation performance
+
+We have benchmarked proof of concept implementation of new logging format. Results measured on 30s benchmark for passing 2 arguments to format function.
+
+|logger                 |time [ns/op]|bytes[B/op]|allocations[alloc/op]|
+|-----------------------|------------|-----------|---------------------|
+|Text Infof             |2252        |248        |3                    |
+|Text InfoS             |2455        |280        |3                    |
+|JSON Infof             |1406        |19         |1                    |
+|JSON InfoS             |319         |67         |1                    |
+
+InfoS implementation for text is 9% slower than Infof. This increase should not have a big impact on overall Kubernetes performance as logging takes less than 2% of overall CPU usage.
+JSON implementation using zapr is much more efficient compared to current implementation.
+
+#### Log volume increase analysis
+
+Example log size change based on apiserver HTTP access log. This log generates most volume in kube-apiserver and accepts lots of arguments. This makes this message a good example of the worst case for log volume increase.
+
+```go
+klog.Infof("%s %s: (%v) %v%v%v [%s %s]", rl.req.Method, rl.req.RequestURI, latency, rl.status, rl.statusStack, rl.addedInfo, rl.req.UserAgent(), rl.req.RemoteAddr)
+```
+
+Using old format generates 206 characters (before [changes by @lavalamp](https://github.com/kubernetes/kubernetes/pull/87203))
+```
+I1025 00:15:15.525108       1 httplog.go:79] GET /api/v1/namespaces/kube-system/pods/metrics-server-v0.3.1-57c75779f-9p8wg: (1.512ms) 200 [pod_nanny/v0.0.0 (linux/amd64) kubernetes/$Format 10.56.1.19:51756]
```

Comment:
Given you touch this already - can someone please finally fix the "$Format" thing? It is clearly a bug - I don't know what it was supposed to be but definitely we didn't want to write "$Format" string.

Reply:
Fixing this could be a great way to reduce log volume increase.